src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -1373,47 +1400,377 @@ public override bool Equals(object obj)
+                    }
+                    var result = remainingTests.Count switch
+                    {
+                        0 => True.Instance,
+                        1 => remainingTests[0],
+                        _ => new AndSequence(remainingTests.ToImmutable()),
+                    };
+                    remainingTests.Free();
+                    return result;
+                }
+                public override BoundDagTest ComputeSelectedTest()
+                {
+                    // Our simple heuristic is to perform the first test of the
+                    // first possible matched case, with two exceptions.
+
+                    if (RemainingTests[0] is One { Test: { Kind: BoundKind.DagNonNullTest } planA })
+                    {
+                        switch (RemainingTests[1])
+                        {
+                            // In the specific case of a null check following by a type test, we skip the
+                            // null check and perform the type test directly.  That's because the type test
+                            // has the side-effect of performing the null check for us.
+                            case One { Test: { Kind: BoundKind.DagTypeTest } planB1 }:
+                                return (planA.Input == planB1.Input) ? planB1 : planA;
+
+                            // In the specific case of a null check following by a value test (which occurs for
+                            // pattern matching a string constant pattern), we skip the
+                            // null check and perform the value test directly.  That's because the value test
+                            // has the side-effect of performing the null check for us.
+                            case One { Test: { Kind: BoundKind.DagValueTest } planB2 }:
+                                return (planA.Input == planB2.Input) ? planB2 : planA;
+                        }
+                    }
+
+                    return RemainingTests[0].ComputeSelectedTest();
+                }
+                public override string Dump(Func<BoundDagTest, String> dump)
                 {
-                    result = Hash.Combine((int)d.Kind, result);
+                    return $"AND({string.Join(", ", RemainingTests.Select(t => t.Dump(dump)))})";
                 }
+            }
 
-                return result;
+            /// <summary>
+            /// A sequence of tests that must be performed, any of which must succeed.
+            /// The sequence is deemed to succeed if some element succeeds.
+            /// </summary>
+            public class OrSequence : SequenceTests
+            {
+                private OrSequence(ImmutableArray<Tests> RemainingTests) : base(RemainingTests) { }
+                public override BoundDagTest ComputeSelectedTest() => this.RemainingTests[0].ComputeSelectedTest();
+                public override Tests Update(ArrayBuilder<Tests> remainingTests) => Create(remainingTests);
```

Comment:
>override Tests Update(ArrayBuilder<Tests> remainingTests) [](start = 23, length = 57)

I can't find any locations where the Update implementation deviates from the Create implementation. Does it make sense to have both methods?

Reply:
This method is virtual, and this override differs from other override (e.g. the override in `AndSequence` inverts the treatment of `True` and `False`, and creates an instance of `AndSequence` instead of `OrSequence`)

---
In reply to: [380975974](https://github.com/dotnet/roslyn/pull/41503#discussion_r380975974) [](ancestors = 380975974)